v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/transforms/SourceWriterFn.java

```diff
@@ -0,0 +1,342 @@
+      LOG.error("Failed to write to source", e);
+    }
+  }
+
+  private com.google.cloud.spanner.Key changeEventToPrimaryKey(
+      String tableName, JsonNode keysJson) {
+    com.google.cloud.spanner.Key.Builder pk = com.google.cloud.spanner.Key.newBuilder();
+
+    try {
+      Table table = ddl.table(tableName);
+      ImmutableList<IndexColumn> keyColumns = table.primaryKeys();
+
+      for (IndexColumn keyColumn : keyColumns) {
+        Column key = table.column(keyColumn.name());
+        Type keyColType = key.type();
+        String keyColName = key.name(); // .toLowerCase();
+        switch (keyColType.getCode()) {
+          case BOOL:
+          case PG_BOOL:
+            pk.append(
+                ChangeEventTypeConvertor.toBoolean(
+                    keysJson, keyColName, /* requiredField= */ true));
+            break;
+          case INT64:
+          case PG_INT8:
+            pk.append(
+                ChangeEventTypeConvertor.toLong(keysJson, keyColName, /* requiredField= */ true));
+            break;
+          case FLOAT64:
+          case PG_FLOAT8:
+            pk.append(
+                ChangeEventTypeConvertor.toDouble(keysJson, keyColName, /* requiredField= */ true));
+            break;
+          case STRING:
+          case PG_VARCHAR:
+          case PG_TEXT:
+            pk.append(
+                ChangeEventTypeConvertor.toString(keysJson, keyColName, /* requiredField= */ true));
+            break;
+          case NUMERIC:
+          case PG_NUMERIC:
+            pk.append(
+                ChangeEventTypeConvertor.toNumericBigDecimal(
+                    keysJson, keyColName, /* requiredField= */ true));
+            break;
+          case JSON:
+          case PG_JSONB:
+            pk.append(
+                ChangeEventTypeConvertor.toString(keysJson, keyColName, /* requiredField= */ true));
+            break;
+          case BYTES:
+          case PG_BYTEA:
+            pk.append(
+                ChangeEventTypeConvertor.toByteArray(
+                    keysJson, keyColName, /* requiredField= */ true));
+            break;
+          case TIMESTAMP:
+          case PG_COMMIT_TIMESTAMP:
+          case PG_TIMESTAMPTZ:
+            pk.append(
+                ChangeEventTypeConvertor.toTimestamp(
+                    keysJson, keyColName, /* requiredField= */ true));
+            break;
+          case DATE:
+          case PG_DATE:
+            pk.append(
+                ChangeEventTypeConvertor.toDate(keysJson, keyColName, /* requiredField= */ true));
+            break;
+          // TODO(b/179070999) -  Add support for other data types.
+          default:
+            throw new IllegalArgumentException(
+                "Column name(" + keyColName + ") has unsupported column type(" + keyColType + ")");
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to update last commit timestamp", e);
+    }
+    return pk.build();
+  }
+
+  private Mutation getShadowTableMutation(
```

Comment:
Can we not use `mutationBuilderFromEvent` from `ChangeEventSpannerConvertor` in spanner-common

Reply:
There is a difference that this one only needs primary key - so I don't really need to create a list of column names and then PK names to call this. The common part is the for loop - added TODO as it needs refactoring for forward as well.